src/constraints/StringConstraints.ts

```diff
@@ -2,8 +2,21 @@ import { ConstraintError } from '../lib/errors/ConstraintError';
 import { Result } from '../lib/Result';
 import type { IConstraint } from './base/IConstraint';
 import { Comparator, eq, ge, gt, le, lt, ne } from './util/operators';
+import net from 'net';
 
-export type StringConstraintName = `s.string.length${'Lt' | 'Le' | 'Gt' | 'Ge' | 'Eq' | 'Ne'}`;
+export type StringConstraintName =
+	| `s.string.${`length${'Lt' | 'Le' | 'Gt' | 'Ge' | 'Eq' | 'Ne'}` | 'regex' | 'url' | 'uuid' | 'email'}`
+	| StringIpName;
+type StringIpName = `s.string.ip${'v4' | 'v6' | ''}`;
+export type StringProtocol = `${string}:`;
+export type StringDomain = `${string}.${string}`;
+export interface UrlOptions {
+	allowedProtocols?: StringProtocol[];
+	allowedDomains?: StringDomain[];
+}
+
+// from https://stackoverflow.com/a/46181/1550155
+export const emailRegex = /^(([^<>()[\]\.,;:\s@\"]+(\.[^<>()[\]\.,;:\s@\"]+)*)|(\".+\"))@(([^<>()[\]\.,;:\s@\"]+\.)+[^<>()[\]\.,;:\s@\"]{2,})$/i;
```

Comment:
Could we maybe use a module for this instead? thoughts @kyranet @favna ?

Reply:
If the module is a fully compliant parser, sure.
If the module is just RegExp#test, then we use the RegExp.

Reply:
We can use 

https://www.npmjs.com/package/email-validator

Reply:
Uhm... [the code](https://github.com/manishsaraan/email-validator/blob/master/index.js) could use a refactor for improved performance, please leave a TODO for me, I'll rewrite it to a faster implementation.